src/GSF.CarbonAware/src/Handlers/EmissionsHandler.cs

```diff
@@ -24,28 +22,89 @@ public EmissionsHandler(ILogger<EmissionsHandler> logger, IEmissionsAggregator a
         _aggregator = aggregator ?? throw new ArgumentNullException(nameof(aggregator));
     }
 
-    /// <inheritdoc />
-    public async Task<double> GetAverageCarbonIntensityAsync(string location, DateTimeOffset start, DateTimeOffset end)
+    ///<inheritdoc/>
+    public async Task<IEnumerable<EmissionsData>> GetEmissionsDataAsync(string location, DateTimeOffset? start = null, DateTimeOffset? end = null)
+    {
+        return await GetEmissionsDataAsync(new string[] { location }, start, end);
+    }
+
+    ///<inheritdoc/>
+    public async Task<IEnumerable<EmissionsData>> GetEmissionsDataAsync(string[] locations, DateTimeOffset? start = null, DateTimeOffset? end = null)
     {
-        var parameters = new CarbonAwareParametersBaseDTO {
+        var parameters = new CarbonAwareParametersBaseDTO
+        {
             Start = start,
             End = end,
-            SingleLocation = location
+            MultipleLocations = locations
         };
-        try {
-            var result = await _aggregator.CalculateAverageCarbonIntensityAsync(parameters);
-            _logger.LogDebug("calculated average carbon intensity: {carbonIntensity}", result);
+        try
+        {
+            var emissionsData = await _aggregator.GetEmissionsDataAsync(parameters);
+            var result = emissionsData.Select(e => (EmissionsData)e);
             return result;
         }
-        catch (Exception ex) when (
-        ex is WattTimeClientException || 
-        ex is WattTimeClientHttpException ||
-        ex is ElectricityMapsClientException ||
-        ex is ElectricityMapsClientHttpException ||
-        ex is LocationConversionException || 
-        ex is ConfigurationException)
+        catch (CarbonAwareException ex)
         {
             throw new Exceptions.CarbonAwareException(ex.Message, ex);
         }
     }
+
+    ///<inheritdoc/>
+    public async Task<IEnumerable<EmissionsData>> GetBestEmissionsDataAsync(string[] locations, DateTimeOffset? start = null, DateTimeOffset? end = null)
+    {
+        var parameters = new CarbonAwareParametersBaseDTO
+        {
+            Start = start,
+            End = end,
+            MultipleLocations = locations
+        };
+        try
+        {
+            var emissionsData = await _aggregator.GetBestEmissionsDataAsync(parameters);
+            var result = emissionsData.Select(e => (EmissionsData)e);
+            return result;
+        }
+        catch (CarbonAwareException ex)
+        {
+            throw new Exceptions.CarbonAwareException(ex.Message, ex);
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<CarbonIntensity> GetAverageCarbonIntensityAsync(string location, TimeRange requestedTimeRange)
+    {
+        return await GetAverageCarbonIntensityAsync(location, new TimeRange[] { requestedTimeRange });
+    }
+    /// <inheritdoc />
+    public async Task<CarbonIntensity> GetAverageCarbonIntensityAsync(string location, TimeRange[] requestedTimeRanges)
```

Comment:
This is a good example of the complexity required to handle multiple values.  We need a new class to talk about start and end, we need to deconstruct it, make single requests, then reconstruct the result in a way the user can tie back to their inputs.  It also now begs the questions, why doesn't this support multiple locations like the others?

It is much easier to add functionality to an SDK than it is to remove it.  I would recommend starting with a smaller footprint instead and waiting for needed overloads to present themselves (rather than guessing).

Reply:
Agreed, will go back to simplified use case and hack away some of the extra models needed